File: recycleradapters/src/main/java/com/rands/lightbulb/recycleradapters/EasyAdapterDiffUtilCallback.java
package com.rands.lightbulb.recycleradapters;

import java.util.List;

import androidx.recyclerview.widget.DiffUtil;

final class EasyAdapterDiffUtilCallback<T extends EasyAdapterDataModel> extends DiffUtil.Callback {
    private final List<T> oldData;
    private final List<T> newData;
    private final EasyAdapterItemsComparator<T> compareCallbacks;

    public EasyAdapterDiffUtilCallback(List<T> oldData, List<T> newData, EasyAdapterItemsComparator<T> compareCallbacks) {
        this.oldData = oldData;
        this.newData = newData;
        this.compareCallbacks = compareCallbacks;
    }

    @Override
    public final int getOldListSize() {
        return oldData.size();
    }

    @Override
    public final int getNewListSize() {
        return newData.size();
    }

    @Override
    public final boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return compareCallbacks.compareItems(oldData.get(oldItemPosition), newData.get(newItemPosition));
    }

    @Override
    public final boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        T oldItem = oldData.get(oldItemPosition);
        T newItem = newData.get(newItemPosition);
        return oldItem.isSelected() == newItem.isSelected() && compareCallbacks.compareItemsContent(oldItem, newItem);
    }
}

File: lightbulb-recycleradapters/src/main/java/com/github/rooneyandshadows/lightbulb/recycleradapters/abstraction/EasyAdapterDiffUtilCallback.kt
package com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction

import androidx.recyclerview.widget.DiffUtil
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.collection.EasyRecyclerAdapterItemsComparator

@JvmSuppressWildcards
internal class EasyAdapterDiffUtilCallback<T : EasyAdapterDataModel>(
    private val oldData: List<SelectableItem<T>>?,
    private val newData: List<SelectableItem<T>>,
    private val compareCallbacks: EasyRecyclerAdapterItemsComparator<T>,
) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldData!!.size
    }

    override fun getNewListSize(): Int {
        return newData.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return compareCallbacks.compareItems(oldData!![oldItemPosition].item, newData[newItemPosition].item)
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldData!![oldItemPosition]
        val newItem = newData[newItemPosition]
        return oldItem.isSelected == newItem.isSelected && compareCallbacks.compareItemsContent(
            oldItem.item,
            newItem.item
        )
    }
}